src/opencl/opencl.cpp

```diff
@@ -0,0 +1,403 @@
+
+        if (instance->platform == NVIDIA)
+            num_command_queues = MAX_NUM_COMMAND_QUEUES;
+        else // Intel
+            num_command_queues = MAX_NUM_COMMAND_QUEUES_INTEL;
+
+        size_t divided_num_pieces = num_pieces / local_size;
+        if (divided_num_pieces == 0) { err = (cl_int)2031; return; } // should never happen since the caller Rust function makes sure that there are more than 1024 pieces
+        while (divided_num_pieces % num_command_queues != 0) {
+            num_command_queues /= 2;
+        }
+        size_t divided_size = next_size / num_command_queues;
+        if (divided_size < 4096 * num_command_queues) { err = (cl_int)2032; return; } // also should never happen
+
+        size_t global_size = num_pieces / num_command_queues;
+
+        err = clEnqueueWriteBuffer(instance->cq[0], ivbuf, CL_TRUE, 0, next_size / 128, (const void*)(iv_offset + copy_offset / 128), 0, NULL, NULL); if (err != CL_SUCCESS) return;
+
+        for (size_t i = 0; i < num_command_queues; i++) {
+            err = clEnqueueWriteBuffer(instance->cq[i], piecebuf, CL_FALSE, i * divided_size, divided_size, (const void*)(inout_offset + divided_size * i + copy_offset), 0, NULL, NULL); if (err != CL_SUCCESS) return;
+        }
+
+        //cl_event event; double duration = 0;
+        for (size_t i = 0; i < num_command_queues; i++) {
+            size_t offset = i * global_size;
+            err = clEnqueueNDRangeKernel(instance->cq[i], instance->kernel, 1, &offset, &global_size, &local_size, 0, NULL, NULL); if (err != CL_SUCCESS) return;
+            //duration += getEventElapsedTime(event, instance->cq[i]);
+        }
+        //printf("Elapsed time for encode kernel: %0.f milliseconds\n", duration / 1000000);
+
+        for (size_t i = 0; i < num_command_queues; i++) {
+            err = clEnqueueReadBuffer(instance->cq[i], piecebuf, CL_FALSE, i * divided_size, divided_size, (void*)(inout_offset + divided_size * i + copy_offset), 0, NULL, NULL); if (err != CL_SUCCESS) return;
+        }
+
+        err = clReleaseMemObject(piecebuf); if (err != CL_SUCCESS) return;
+        err = clReleaseMemObject(ivbuf); if (err != CL_SUCCESS) return;
+
+        remaining_size -= next_size;
+    }
+
+    err = clFlush(instance->cq[num_command_queues - 1]); if (err != CL_SUCCESS) return;
+    err = clFinish(instance->cq[num_command_queues - 1]); if (err != CL_SUCCESS) return;
+
+    //auto end = std::chrono::high_resolution_clock::now();
+    //std::cout << "Total elapsed time for GPU encode on platform " << (int)instance->platform << ": " << (std::chrono::duration<double>(end - start)).count() << " seconds" << std::endl;
+}
+
+#define CPU_FACTOR 0.75f
+#define NVIDIA_FACTOR 100.0f
+#define AMD_FACTOR 15.0f
+#define INTEL_FACTOR 1.3f
+
+extern "C"
+cl_int sloth256_189_opencl_batch_encode(unsigned char* inout, size_t total_size, unsigned char* iv, size_t layers, EncodeOpenCLInstances* instances) {

```

Comment:
Explanation throughout this function is needed, I don't fully understand what is happening here.